src/build/__init__.py

```diff
@@ -201,15 +187,20 @@ def check_dependencies(self, distribution):  # type: (str) -> Set[str]
 
         return {dep for dep in dependencies if not check_version(dep)}
 
-    def build(self, distribution, outdir):  # type: (str, str) -> None
+    def build(self, distribution, outdir=None):  # type: (str, Optional[str]) -> None
         """
         Builds a distribution
 
         :param distribution: Distribution to build (sdist or wheel)
         :param outdir: Output directory
         """
         build = getattr(self.hook, 'build_{}'.format(distribution))
-        outdir = os.path.abspath(outdir)
+
+        if outdir is None:
+            # resolve outdir relative to project root if not specified.
+            outdir = os.path.join(self.srcdir, 'dist')
```

Comment:
Is that showing what happens or what you expect to happen?

Reply:
What I expect to happen.

Reply:
Well, I implemented it as requested in #69:

> The default dist directory should be located in the project root no matter what the current working directory is; a custom destination directory should be located relative to the current working directory, no matter where the project is.

Since nobody said otherwise I assumed this is what you wanted ;)

Reply:
I agree here with @layday here. Historically upon no destination specified the target was always within the dist folder of the source directory. So my expectation would be that by default it populates it there...

Reply:
From my experience dealing with different build systems and tools, I disagree with that, it usually tends to give trouble. @uranusjr can you expand here?

PEP 517 makes no commentary on this, so it's up to us to choose how to implement it.

Even if we wanted this behavior, this would be the wrong place to do it. We should calculate the absolute path in the main and set it as default.

```
  --outdir dist, -o dist
                        output directory
```

I absolutely expect that to create a `dist` directory in the current path.

Reply:
I'm kinda torn here, because my intuition is that if I'm executing `build`, the output directory would be relative to CWD, because it feels weird to modify the project source tree.

On the other hand, the use case in https://github.com/pypa/build/issues/69#issuecomment-704312424 modifies the source tree _anyway_, it just does it in some random nested location, which is not what I think people would want for the most part.

I'm tempted to say that the default should be, "Project root if CWD is a subdirectory of the project root, otherwise CWD", but I think that will give people false intuitions about how `build` works ☹.

I guess I have switched to +0 on "always in project root" because there are many situations where writing out the location of the project root is less than simple (e.g. `build ~/very/long/path` → `build ~/very/long/path -o ~/very/long/path/dist`), whereas it is _always_ easy to refer to the CWD, so `build ~/very/long/path` → `build ~/very/long/path -o ./dist`. When you have two essentially equally reasonable defaults A and B and getting behavior A when behavior B is the default is easier than getting behavior B when behavior A is the default, you should make behavior B the default.

Reply:
> Historically upon no destination specified the target was always within the dist folder of the source directory.

No? Where does this historically come from? I consider setuptools to be the historic approach and it does not support building projects from different directories, that might break things as it assumes `setup.py` is being invoked from the project root. But even if we do that.

```
$ ls
setup.py test.py
$ cat setup.py
import setuptools

setuptools.setup(name='test', version='1.0.0')
$ mkdir somedir
$ cd somedir
$ python ../setup.py sdist
running sdist
running egg_info
creating test.egg-info
writing test.egg-info/PKG-INFO
writing dependency_links to test.egg-info/dependency_links.txt
writing top-level names to test.egg-info/top_level.txt
writing manifest file 'test.egg-info/SOURCES.txt'
reading manifest file 'test.egg-info/SOURCES.txt'
writing manifest file 'test.egg-info/SOURCES.txt'
warning: sdist: standard file not found: should have one of README, README.rst, README.txt, README.md

running check
warning: Check: missing required meta-data: url

warning: Check: missing meta-data: either (author and author_email) or (maintainer and maintainer_email) must be supplied

creating test-1.0.0
creating test-1.0.0/test.egg-info
copying files to test-1.0.0...
copying test.egg-info/PKG-INFO -> test-1.0.0/test.egg-info
copying test.egg-info/SOURCES.txt -> test-1.0.0/test.egg-info
copying test.egg-info/dependency_links.txt -> test-1.0.0/test.egg-info
copying test.egg-info/top_level.txt -> test-1.0.0/test.egg-info
Writing test-1.0.0/setup.cfg
creating dist
Creating tar archive
removing 'test-1.0.0' (and everything under it)
$ ls dist
test-1.0.0.tar.gz
$ cd ..
$ tree
.
├── setup.py
├── somedir
│   ├── dist
│   │   └── test-1.0.0.tar.gz
│   └── test.egg-info
│       ├── dependency_links.txt
│       ├── PKG-INFO
│       ├── SOURCES.txt
│       └── top_level.txt
└── test.py

3 directories, 7 files
```

Reply:
On the topic of the 'metavar', I expect it to describe the input, not to provide a default.  I find 'dist' in the help text to be misplaced.

Reply:
> > Historically upon no destination specified the target was always within the dist folder of the source directory.
> 
> No? Where does this historically come from?

Personally I think we shouldn't worry about the historical record here. `build` is a new tool, users shouldn't expect it to have the same defaults as other tools, just that it has reasonable defaults.

Reply:
> I guess I have switched to +0 on "always in project root" because there are many situations where writing out the location of the project root is less than simple (e.g. `build ~/very/long/path` → `build ~/very/long/path -o ~/very/long/path/dist`), whereas it is _always_ easy to refer to the CWD, so `build ~/very/long/path` → `build ~/very/long/path -o ./dist`. 

I mean
```
$ python -m build -o ~/very/long/path{/dist,}
```
But sure. You also have ctrl+c ctrl+v.

> When you have two essentially equally reasonable defaults A and B and getting behavior A when behavior B is the default is easier than getting behavior B when behavior A is the default, you should make behavior B the default.

What if project root is read only? What if the project happens to contain a `dist` directory? There is also the breaking of the common build system default have the outdir relative to the cwd.
So, I don't find this equally reasonable. The amount of people building from a different path is already pretty small, I think it would be best to behave like build systems commonly do.

Reply:
> What if project root is read only? What if the project happens to contain a `dist` directory?

These questions sound to me just as valid for CWD. What if I'm already in a project directory, and I want to build for another project?

Reply:
Well, they are, but they are slightly better there.

Regarding the read only issue, it is fairly unlikely and if it happens it will probably be in a custom build environment. In custom build environments it would be much more likely that the source is ready only rather than the current directory.

Regarding the `dist` directory, I'd actually argue this might be a useful use case.

```
$ python -m build project1
$ python -m build project2
$ python -m build project3
$ python -m build project4
$ python -m build project5
$ ls dist
project1-1.0.0-py2.py3-none-any.whl  project1-1.0.0.tar.gz
project1-1.0.0-py2.py3-none-any.whl  project1-1.0.0.tar.gz
project1-1.0.0-py2.py3-none-any.whl  project1-1.0.0.tar.gz

Reply:
But my main problem here is really us behaving differently than what I would expect. All these other cases are pretty unlikely and shouldn't matter that much.

Reply:
> That could be a bit annoying, but I usually use `dist` directories as throw away and I suspect most people do too.

I don't, but maybe it's just me. I'm +0 in between them, so 🤷

Reply:
Do you want this handled in ProjectBuilder or the CLI?  You mentioned:

> Even if we wanted this behavior, this would be the wrong place to do it.

Reply:
> What if project root is read only? What if the project happens to contain a `dist` directory? There is also the breaking of the common build system default have the outdir relative to the cwd.

Then the build will fail with a warning?

We're not saying what people can and can't do, just trying to decide on a reasonable default. I think "dist goes in the project root" and "dist goes in CWD" are pretty much equally valid — when explicitly asked about this, I thought it should be CWD, but I now recall that at some point fairly recently I was bitten by this because CWD was a directory like `tests`, and I executed `python -m build ..` only to have a `dist` directory dumped in a random subdir (which was also not in my `.gitignore`...), so obviously I assumed that the behavior was that the default value for `-o` was `$PROJECT_ROOT/dist`.


> ```
> $ python -m build -o ~/very/long/path{/dist,}
> ```
> 
> But sure. You also have ctrl+c ctrl+v.

I know about shell expansion, but if you didn't explicitly tell me that `~/very/long/path{/dist,}` would pass one argument to `-o` and one argument to `build`, I would be very uncertain about it. Ctrl + v Ctrl + c is also more of a pain. Obviously there are lots of ways to accomplish this, and we're just talking about defaults, but it seems pretty clear to me that at best it's a toss-up whether the assumption is "default goes in the project root" or "default goes in CWD" (with more people weighing in that they assume it's project root), _and_ the way to get the "put output in CWD" behavior is way simpler than the way to get the "put output in project root" (assuming the other one is the default), which makes me think we should make "output in project root" the default.

Reply:
I'd personally leave the code changes as is, just use getcwd instead of srcdir when calculating the path 🤔

Reply:
If we do change the current behavior, it should be in `__main__.py` , not `ProjectBuilder`, yes.

Reply:
Why? I'd keep our programmatic API and UX in sync so both should be defined within project builder.

Reply:
My experience from various build tools taught me that the build tool should build to a predicatable location relative to the project root by default, and redicatablly resolve to cwd if a relative path is manually provided. But now that I think about this through again, the behaviour can actually be interpreted either way, since *most of those tools do not actually distinguish between the project root and cwd*. GNU Make, for example, explicitly defines `-C` to mean *change cwd to this directory*, so cwd is always the same as the project root.

The only build tool that actually distinguish between cwd and project root, from the top of my head, is Cargo; `cargo build --manifest-path=path/to/source --target-dir=path/to/target` would resolve `target-dir` against cwd, not the project root (where `Cargo.toml` locates).

Still, though, it feels extremely weird for me that `build` would generate a `dist` directory in cwd by default, when the target directory is not explicitly specified, since I expect this “directory that holds artifacts” thing to be only generated in the project root, not in the directory “I’m currently in” since that’s most tools end up doing, no matter if they do it by resolving explicitly (like Cargo) or changing the cwd (like everyone else). This is kind of irrational, since I probably wouldn’t mind if it generates the files directly in cwd without nesting them in the `dist` directory (CMake generates the Makefile directly inside cwd when you build out-of-source). But still, the current behaviour is quite surprising to me based on my own experience.

Reply:
> Why? I'd keep our programmatic API and UX in sync so both should be defined within project builder.

Well, I'd revert the changes to make this arguments optional.

If we do make it have this behavior, `python -m build --help` needs to be absolutely clear where it is placing the files.

Reply:
It is even now we added it to the help message, see above.

Reply:
There might be an argument to be made that the programmatic API should not mirror the CLI and that neither should the project root default to the working directory in `ProjectBuilder`. I can think of two reasons:

1. The working directory is more difficult to discern outside of the terminal.
1. `ProjectBuilder` might not be typically invoked in or relative to the working directory, e.g. if it were to be adopted by pip to build downloaded sdists.  It will then become something of a gotcha for pip maintainers.

Reply:
I've thrown out a quick twitter poll (https://twitter.com/gjbernat/status/1319724762969636867?s=19) and 14 against 11 people prefer to default to the predictable location within the project folder as @uranusjr says. This show to me this is indeed a coin toss, but maybe a slight more towards the proposal of this PR than the status quo. Following this I change my vote to +1 for this PR.

Reply:
> There might be an argument to be made that the programmatic API should not mirror the CLI and that neither should the project root default to the working directory in `ProjectBuilder`. I can think of two reasons:
> 
> 1. The working directory is more difficult to discern outside of the terminal.
> 1. `ProjectBuilder` might not be typically invoked in or relative to the working directory, e.g. if it were to be adopted by pip to build downloaded sdists.  It will then become something of a gotcha for pip maintainers.

I'd be surprised if pip does not override the default with explicit absolute paths 🤔

Reply:
I am fine with changing the default output directory to under the project root, but I would really like to keep the API the same. I don't think we should be depending on cwd for anything there and API users should be more than capabale, in fact I would they say most would actually prefer, to specify the paths themselves. IMO we should as little opinionated as possible in the API, and having this sort of defaults is unecessary.

Reply:
> I am fine with changing the default output directory to under the project root, but I would really like to keep the API the same. I don't think we should be depending on cwd for anything there and API users should be more than capable, in fact I would they say most would actually prefer, to specify the paths themselves. IMO we should as little opinionated as possible in the API, and having this sort of defaults is unnecessary.

I'm on board with this. For the CLI there pretty much must be a default, but we have two strong options. For people using the programmatic API, the calculus is entirely different — the feedback loops are probably longer, they are likely passing along variables rather than string literals, etc, so I think it makes sense to go with "no default in `ProjectBuilder`, default to project root in CLI".

Reply:
@layday, can you update the PR to adhere to the changes requested?

To summarize, we want to keep the API as is and change the default value for the output directory. I think changing the default value should come in a different PR though, keeping this one to fixing the `outdir` resolution.

Reply:
`ProjectBuilder` would default to `'.'` for the srcdir prior to this PR - if there is no objection I will remove the defaults from `__init__` and `build`.

Reply:
Yeah, please remove all optional paths :blush:

Reply:
Unless I am misunderstanding the only thing left to do here is to remove the default value from the `ProjectBuilder`'s `srcdir` if we're introducing the project root path change in a separate PR.

Reply:
Things got a little muddled here so I've opened #160 for the API change.  Looking at #69 again I am not able to reproduce the issue @uranusjr was having with `outdir` being created relative to the `srcdir` and at this point I'm not really sure if I was able to reproduce it originally either!